platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,927 @@
+- Render the child `Page`s
+
+##### `Page` Component
+**Purpose:**
+A `Page` is used to define the path for a page and render its form contents and
+the normal page chrome.
+
+**Usage requirements:**
+A `Page` may be the child of either `Router` directly or `Chapter`.
+
+**Responsibilities:**
+- If no `overrideFormik` prop is passed
+  - Render a [`<Formik>`](https://formik.org/docs/api/formik) component and pass
+    it some default props for `initialValues`, `validate`, and `onSubmit`
+  - If the save-in-progress React Context is available, `Page` will use it to
+    auto-save the form
+- Render the children, passing them the extra props for form state
+  - Extra props include `formData` and `pathParams`
+  - For an example usage, see [below](#functionality-array-pages)
+- Render the form navigation buttons "Back" and "Continue"
+- Prevent navigation if there are validation errors
+
+**A note on cloning children**
+To pass the form state to `Page`'s children, it will clone all its children and
+pass the form state props along with the child's given props. Implementation
+will look something like the following:
+```js
+const newChildren = React.Children.map(
+  this.props.children,
+  child => React.cloneElement(child, { formData, pathParams }));
+```
+
+This means developers can create form pages in a couple different ways. 
+
+_Using the form state:_
+```jsx
+<Page>
+  {({ formData, pathParams }) => (/* Use the form state to render the correct fields */)}
+</Page
+```
+
+_Not using the form state:_
+```jsx
+<Page>
+  <StringField name="foo" />
+  <StringField name="bar" />
+</Page
+```
+
+In the above example, both `StringField`s will have the following props:
+- `name`
+- `formData`
+- `pathParams`
+
+The form state props `formData` and `pathParams` will simply go unused.
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
+
+**Usage requirements:**
+- `Route` may only be a child of `Router`
+
+**Why not just use `Page`?**
+The `Chapter` and `Page` components are meant to be used to define pages where
+the user may enter data. These pages are sequential and have back and continue
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <FieldArray
+      name="marriages"
+      render={arrayHelpers => (
+        <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+      )}
```

Comment:
It isn't clear what `arrayHelpers` is doing in this function.

Reply:
This [comes from Formik](https://formik.org/docs/api/fieldarray#fieldarray-helpers). In this function, it isn't doing anything. I opted to keep the example simple and not use it. :man_shrugging: 

I'll either add a link to the `FieldArray` Formik documentation to clarify the purpose, or simplify the example. :+1: